File: Algorithm/greedy algorithm/atcoder/B39/B39.py
# 以下に、**Python (CPython 3.11.4)** を用いた解答を提示します。
# この解法は **優先度付きキュー（最大ヒープ）+ 日付ごとの仕事分類** を利用し、\*\*最も効率的な満点解法（O(N log N + D log N))）\*\*です。

# ---

# ## ✅ 解法の要点（Python）

# * `heapq` はデフォルトで **最小ヒープ** なので、**負の値を使って最大ヒープのように扱います**。
# * 各日 `d` に対し、その日に開始できる仕事の報酬をヒープに追加。
# * 毎日、ヒープから最大報酬の仕事を1つ選んで合計。

# ---

# ## ✅ 実装コード（関数化・型注釈付き）

# ```python
import sys
import heapq
from typing import List, Tuple


def get_max_earnings(n: int, d: int, jobs: List[Tuple[int, int]]) -> int:
    """
    最大収益を計算する関数

    Parameters:
    - n: int - 仕事の数
    - d: int - 就業可能日数
    - jobs: List[Tuple[int, int]] - 各仕事の情報 (開始日Xi, 報酬Yi)

    Returns:
    - int - 得られる最大報酬の合計
    """

    # 各日ごとの仕事リストに分類
    job_by_day: List[List[int]] = [[] for _ in range(d + 1)]
    for x, y in jobs:
        if x <= d:
            job_by_day[x].append(y)

    # 最大ヒープ（heapqで負値にして使用）
    max_heap: List[int] = []
    total: int = 0

    for day in range(1, d + 1):
        # 今日から可能な仕事をヒープに追加
        for reward in job_by_day[day]:
            heapq.heappush(max_heap, -reward)  # 最大ヒープのためマイナス

        # 今日やるべき最大報酬の仕事を選択
        if max_heap:
            total += -heapq.heappop(max_heap)

    return total


def main() -> None:
    """
    標準入力からデータを受け取り、最大収益を出力するメイン関数
    """

    input_data = sys.stdin.read().strip().split('\n')
    n, d = map(int, input_data[0].split())
    # jobs: List[Tuple[int, int]] = [tuple(map(int, line.split())) for line in input_data[1:]]
    # ✅ 型安全でpylanceエラーを回避
    jobs: List[Tuple[int, int]] = [
        (int(x), int(y)) for x, y in (line.split() for line in input_data[1:])
    ]

    result: int = get_max_earnings(n, d, jobs)
    print(result)


if __name__ == "__main__":
    main()
# ```

# ---

# ## ✅ 使用例（テストケース）

# 入力:

# ```
# 5 4
# 1 1
# 2 4
# 2 3
# 3 4
# 4 2
# ```

# 出力:

# ```
# 12
# ```

# ---

# ## ✅ 時間・メモリ効率分析

# | 処理    | 計算量                  | 備考                  |
# | ----- | -------------------- | ------------------- |
# | job分類 | O(N)                 | 各仕事を1回だけ処理          |
# | ヒープ操作 | O(N log N + D log N) | push/pop 1回あたり logN |
# | 合計メモリ | O(N + D)             | job\_by\_day + ヒープ  |
# | 最大入力  | N=200,000, D=2000    | 問題制約内で高速に処理可能       |

# ---

# ## ✅ 特長

# * ヒープ（heapq）で最大値を管理し、**毎日最適な仕事を選ぶ貪欲戦略**
# * Pythonでも**満点が取れる効率的実装**
# * `typing` により明確な型指定、可読性・保守性も考慮

# ---

# ご希望があればこのコードの**処理過程のステップ解説や図解付きの説明**も対応可能です。お気軽にどうぞ。

# このエラーは、`Pylance`（VSCodeなどで使われている型チェッカー）が `List[Tuple[int, int]]` という期待される型と、`list[tuple[int, ...]]`（要素数が2に限られていないタプル）との**不一致**を検出していることによるものです。

# ---

# ## ✅ 原因

# 次のようなコードで `jobs` を生成している部分：

# ```python
# jobs: List[Tuple[int, int]] = [tuple(map(int, line.split())) for line in input_data[1:]]
# ```

# ここで `tuple(map(...))` を使うと、Pylance には `Tuple[int, ...]`（要素数が可変）と解釈され、`Tuple[int, int]`（要素数2）とは型不一致と判断されます。

# ---

# ## ✅ 解決策（安全で型正しい書き換え）

# `map(int, ...)` から明示的に2要素のタプルを作るように書き換えましょう：

# ```python
# jobs: List[Tuple[int, int]] = [ (int(x), int(y)) for x, y in (line.split() for line in input_data[1:]) ]
# ```

# ---

# ## ✅ 修正後のコード抜粋

# ```python
# def main() -> None:
#     """
#     標準入力からデータを受け取り、最大収益を出力するメイン関数
#     """

#     input_data = sys.stdin.read().strip().split('\n')
#     n, d = map(int, input_data[0].split())

#     # ✅ 型安全でpylanceエラーを回避
#     jobs: List[Tuple[int, int]] = [
#         (int(x), int(y)) for x, y in (line.split() for line in input_data[1:])
#     ]

#     result: int = get_max_earnings(n, d, jobs)
#     print(result)
# ```

# ---

# ## ✅ 解説

# | 書き方                    | 型推論               | 説明        |
# | ---------------------- | ----------------- | --------- |
# | `tuple(map(int, ...))` | `tuple[int, ...]` | 任意長、要素数不定 |
# | `(int(x), int(y))`     | `Tuple[int, int]` | 要素数2で型一致  |

# ---

# これで `Pylance` の型エラーは解消されます。
# 他にも型エラーや警告が出る場合はお気軽にお知らせください。

File: Algorithm/greedy algorithm/atcoder/B39/B39.js
// 以下は、Node.js (`fs` モジュール使用, Node.js 18.16.1 準拠) を用いて、**最大収入を得るスケジューリング問題**を効率的に解くコードです。

// この問題は、`D日間` で `X_i日目以降に可能な仕事を最大で1日1件` という条件の下で、**貪欲法 + 優先度付きキュー（ヒープ）** を用いて最大収益を得る典型問題です（鉄則本 8.3 節に該当）。

// ---

// ### ✅ 解法概要（満点解法）：

// * 各日 `d` について、できる仕事候補（X\_i ≤ d）を収集しておき、`最大の報酬Y_i` の仕事を選ぶ。
// * 実装には **最大ヒープ** を使って毎日最も報酬の高い仕事を選ぶ。
// * 時間計算量：`O(N log N + D log N)` → 満点解法

// ---

// ### ✅ 実装コード（Node.js + fs + 最大ヒープ）：

// ```javascript
const fs = require('fs');

/**
 * 最大収益を求める関数
 * @param {number} N - 仕事の数
 * @param {number} D - 就業可能日数
 * @param {Array<[number, number]>} jobs - 仕事一覧（[開始日, 報酬]）
 * @returns {number} - 最大収益
 */
function getMaxEarnings(N, D, jobs) {
    // 各日の仕事を格納（day: 1〜D）
    const jobByDay = Array.from({ length: D + 1 }, () => []);
    for (const [x, y] of jobs) {
        if (x <= D) jobByDay[x].push(y);
    }

    // 最大ヒープ（優先度付きキュー）
    const maxHeap = new MaxHeap();
    let total = 0;

    for (let day = 1; day <= D; day++) {
        // 今日できる仕事をヒープに追加
        for (const reward of jobByDay[day]) {
            maxHeap.push(reward);
        }
        // 今日の最良の仕事を選ぶ
        if (maxHeap.size() > 0) {
            total += maxHeap.pop();
        }
    }

    return total;
}

// --- ヒープ実装（最大ヒープ） ---
class MaxHeap {
    constructor() {
        this.data = [];
    }

    push(val) {
        this.data.push(val);
        this._heapifyUp(this.data.length - 1);
    }

    pop() {
        if (this.data.length === 0) return null;
        const top = this.data[0];
        const last = this.data.pop();
        if (this.data.length > 0) {
            this.data[0] = last;
            this._heapifyDown(0);
        }
        return top;
    }

    size() {
        return this.data.length;
    }

    _heapifyUp(i) {
        while (i > 0) {
            const parent = Math.floor((i - 1) / 2);
            if (this.data[parent] >= this.data[i]) break;
            [this.data[parent], this.data[i]] = [this.data[i], this.data[parent]];
            i = parent;
        }
    }

    _heapifyDown(i) {
        const n = this.data.length;
        while (true) {
            let largest = i;
            const left = 2 * i + 1;
            const right = 2 * i + 2;

            if (left < n && this.data[left] > this.data[largest]) largest = left;
            if (right < n && this.data[right] > this.data[largest]) largest = right;

            if (largest === i) break;

            [this.data[i], this.data[largest]] = [this.data[largest], this.data[i]];
            i = largest;
        }
    }
}

// --- 入出力処理 ---
function main() {
    const input = fs.readFileSync('/dev/stdin', 'utf8').trim().split('\n');
    const [N, D] = input[0].split(' ').map(Number);
    const jobs = input.slice(1).map((line) => line.split(' ').map(Number));
    const result = getMaxEarnings(N, D, jobs);
    console.log(result);
}

main();

// ### ✅ 実行性能（目安）

// * 時間計算量：`O(N log N + D log N)`（`N ≦ 2e5`, `D ≦ 2000`）
// * 空間計算量：`O(N + D)`（jobByDayとヒープ分）

// ---

// ### ✅ 入力例テスト（標準入力）

// ```
// 5 4
// 1 1
// 2 4
// 2 3
// 3 4
// 4 2
// ```

// ```
// 出力:
// 12
// ```

// ---

// ご希望があれば、このコードの**図解付きステップ解説**も提供可能です。
